tools/repl/repl.cpp

```diff
@@ -222,6 +228,36 @@ static bool needsAnotherLine(llvm::StringRef input) {
   return !stack.empty();
 }
 
+#if HAVE_LIBREADLINE
+// Load history file or create it
+std::error_code loadOrCreateHistoryFile(llvm::SmallString<128>& historyFile) {
+  if (!llvm::sys::path::home_directory(historyFile)) {
+    // Use ENOENT here since it could not found a home directory
+    return std::error_code(ENOENT, std::system_category());
+  }
+
+  llvm::sys::path::append(historyFile, historyFileBaseName);
+
+  if (!llvm::sys::fs::exists(historyFile)) {
+    int fd;
+    auto err = llvm::sys::fs::openFileForWrite(llvm::Twine(historyFile), fd);
```

Comment:
I'm running into some test errors around this code.  This creates an empty file, which causes a confusing error from `::read_history()`:
```
$ rm ~/.hermes_history
$ ./bin/hermes-repl
Could not load history file: No such file or directory
>>
```
I did some experimentation, and this seems to be due to the file being empty, and not including the magic version header.  If I start with `cp /dev/null ~/.hermes_history` I get the same warning.

And if I C-d immediately, so that a file is written with no contents, then when I run the next time, I get a different confusing error:
```
$ cat ~/.hermes_history
_HiStOrY_V2_
$ ./bin/hermes-repl
Could not load history file: Operation not permitted
>>
```
What's the purpose of creating the empty file here?  Can we suppress the empty write, so we don't get a warning?

Finally, I noticed the history file is mode 600, even though my umask is 0022.  Shouldn't it be mode 644?  Although I see other history files with the same mode, so maybe this is a readline "feature", it's not universal: node's history file is mode 644.

Sorry for all this.  The libreadline history support seems unfortunately brittle.

Reply:
Sometimes it could happens when you don't have a `HOME` environment variable set. Besides that, I really don't know what could happens.

The empty file is created in order to `write_history` to work, it opens the history file only on "append" mode, so if there is no file it returns a `ENOENT` error.

The LLVM file system API should actually create a 666 mode, but 600 is actually a good choice because of security issues on accessing history files.

I'm reading the libreadline source code in order to understand it.

Reply:
So, I'm avoiding creating empty files and ignoring `ENOENT` errors, because it happens in this case and that's an expected behavior.

About the permission, actually there is nothing we could do about it (at least using libreadline), both version [7](http://git.savannah.gnu.org/cgit/readline.git/tree/histfile.c?id=6c32f81cd66bbe86218469063690c84205661a5e#n649) and [8](http://git.savannah.gnu.org/cgit/readline.git/tree/histfile.c?id=8e6ccd0373d77b86ed37a9a7d232ccfea3d6670c#n654) set the file permission hardcoded.